A47-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,181 @@
+A47: xDS Least Request LB Policy
+----
+* Author(s): erikjoh & [fabric](https://github.com/orgs/spotify/teams/fabric)
+* Approver: a11r
+* Status: Implementation in progress
+* Implemented in: Java in progress
+* Last updated: 2021-10-18
+* Discussion at: TBD
+
+## Abstract
+
+Add support for least request load balancing configured via xDS.
+
+## Background
+
+In multi-tenancy environments like for example Kubernetes, some upstream endpoints may perform much slower than others.
+When using plain round-robin load balancing, this becomes a problem since all endpoints get an equal amount of RPCs
+even when some endpoints perform worse. The result is increased tail latencies and in extreme cases even increased error rates.
+As an alternative to round-robin, Envoy provides a "least request" or "weighted least request"
+[load balancing policy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#weighted-least-request)
+which uses per-endpoint outstanding/active request counters as a heuristic for determining load on each endpoint.
+
+### Related Proposals: 
+
+This proposal builds on earlier work described in the following gRFCs:
+* [gRFC A27: xDS-based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+* [gRFC A42: xDS Ring Hash LB Policy](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md)
+
+## Proposal
+
+Support `LEAST_REQUEST` LB policy and configuration of it using xDS.
+Specifically, gRPC will support the "*all weights equal*" least request algorithm as
+[defined by Envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#weighted-least-request).
+
+### Least Request LB Policy
+
+#### xDS API Fields
+
+The xDS `Cluster` resource specifies the load balancing policy to use
+when choosing the endpoint to which each request is sent.  The policy to
+use is specified in the [`lb_policy`
+field](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L722).
+Prior to this proposal, gRPC supported two values for this field,
+`ROUND_ROBIN` and `RING_HASH`.  With this proposal, we will add support for an
+additional policy, `LEAST_REQUEST`.
```

Comment:
Should this support a list of load balancing algorithms as what google suggested in the previous meeting?